packages/test/src/test-integration-update.ts

```diff
@@ -0,0 +1,259 @@
+    t.deepEqual(wfResult, ['fail-update', 'done', '$']);
+  });
+});
+
+test('Update is rejected if there is no handler', async (t) => {
+  const { createWorker, startWorkflow, assertWorkflowUpdateFailed } = helpers(t);
+  const worker = await createWorker();
+  const updateWithoutHandler = wf.defineUpdate<string[], [string]>('updateWithoutHandler');
+  await worker.runUntil(async () => {
+    const wfHandle = await startWorkflow(workflowWithUpdates);
+    await assertWorkflowUpdateFailed(
+      wfHandle.executeUpdate(updateWithoutHandler, { args: [''] }),
+      wf.ApplicationFailure,
+      'Update has no handler: updateWithoutHandler'
+    );
+  });
+});
+
+test('Update sent after workflow completed', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+  const worker = await createWorker();
+  await worker.runUntil(async () => {
+    const wfHandle = await startWorkflow(workflowWithUpdates);
+    await wfHandle.executeUpdate(doneUpdate);
+    await wfHandle.result();
+    try {
+      await wfHandle.executeUpdate(update, { args: ['1'] });
+    } catch (err) {
+      t.true(err instanceof wf.WorkflowNotFoundError);
+      t.is((err as wf.WorkflowNotFoundError).message, 'workflow execution already completed');
+    }
+  });
+});
+
+test('Update id can be assigned and is present on returned handle', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+  const worker = await createWorker();
+  await worker.runUntil(async () => {
+    const wfHandle = await startWorkflow(workflowWithUpdates);
+    const updateHandle = await wfHandle.startUpdate(doneUpdate, { updateId: 'my-update-id' });
+    t.is(updateHandle.updateId, 'my-update-id');
+  });
+});
+
+// The following tests construct scenarios in which doUpdate jobs are packaged
+// together with startWorkflow in the first Activation. We test this because it
+// provides test coverage for Update buffering: were it not for the buffering,
+// we would attempt to start performing the Update (validate and handle) before
+// its handler is set, since sdk-core sorts Update jobs with Signal jobs, i.e.
+// ahead of jobs such as startWorkflow and completeActivity that might result in
+// a setHandler call. Also note that we do need to make this guarantee to users,
+// because a user might know that their Update is in the first WFT, for example
+// because they are doing something similar to what this test does to achieve
+// that.
+
+// TODO: we currently lack a way to ensure, without race conditions, via SDK
+// APIs, that Updates are packaged together with startWorkflow in the first
+// Activation. In lieu of a non-racy implementation, the test below does the
+// following:
+// 1. Client sends and awaits startWorkflow.
+// 2. Client sends but does not await executeUpdate.
+// 3. Wait for long enough to be confident that the server handled the
+//    executeUpdate and is now waiting for the Update to advance to Completed.
+// 4. Start the Worker.
```

Comment:
This branch contains a fix for the race condition here: https://github.com/dandavison/temporalio-sdk-typescript/pull/1

It uses the recently-added non-blocking Update poll. This required the `sdk-core` protos to be updated, and so is blocked on this `sdk-core` PR: https://github.com/temporalio/sdk-core/pull/647.